src/include/OpenImageIO/hash.h

```diff
@@ -59,6 +59,172 @@ OIIO_NAMESPACE_BEGIN
 using std::unordered_map;
 using std::hash;
 
+namespace fasthash {
+
+/* The MIT License
+   Copyright (C) 2012 Zilong Tan (eric.zltan@gmail.com)
+   Permission is hereby granted, free of charge, to any person
+   obtaining a copy of this software and associated documentation
+   files (the "Software"), to deal in the Software without
+   restriction, including without limitation the rights to use, copy,
+   modify, merge, publish, distribute, sublicense, and/or sell copies
+   of the Software, and to permit persons to whom the Software is
+   furnished to do so, subject to the following conditions:
+   The above copyright notice and this permission notice shall be
+   included in all copies or substantial portions of the Software.
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+   NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+   BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+   ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+   CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+   SOFTWARE.
+*/
+
+// Compression function for Merkle-Damgard construction.
+// This function is generated using the framework provided.
+static inline uint64_t mix(uint64_t h) {
+	h ^= h >> 23;
+	h *= 0x2127599bf4325c37ULL;
+	h ^= h >> 47;
+	return h;
+}
+
+inline uint64_t fasthash64(const void *buf, size_t len, uint64_t seed)
+{
+	const uint64_t    m = 0x880355f21e6d1965ULL;
+	const uint64_t *pos = (const uint64_t *)buf;
+	const uint64_t *end = pos + (len / 8);
+	const unsigned char *pos2;
+	uint64_t h = seed ^ (len * m);
+	uint64_t v;
+
+	while (pos != end) {
+		v  = *pos++;
+		h ^= mix(v);
+		h *= m;
+	}
+
+	pos2 = (const unsigned char*)pos;
+	v = 0;
+
+	switch (len & 7) {
+        case 7: v ^= (uint64_t)pos2[6] << 48;
+        case 6: v ^= (uint64_t)pos2[5] << 40;
+        case 5: v ^= (uint64_t)pos2[4] << 32;
+        case 4: v ^= (uint64_t)pos2[3] << 24;
+        case 3: v ^= (uint64_t)pos2[2] << 16;
+        case 2: v ^= (uint64_t)pos2[1] << 8;
+        case 1: v ^= (uint64_t)pos2[0];
+                h ^= mix(v);
+                h *= m;
+	}
+
+	return mix(h);
+}
+
+// simplified version for hashing just a few ints
+inline uint64_t fasthash64(const std::initializer_list<uint64_t> buf) {
+	const uint64_t    m = 0x880355f21e6d1965ULL;
+	uint64_t h = (buf.size() * sizeof(uint64_t)) * m;
+	for (const uint64_t v : buf) {
+		h ^= mix(v);
+		h *= m;
+	}
+	return mix(h);
+}
+
+} // namespace fasthash
+
+#ifdef __AES__
```

Comment:
The way you're done this, the function itself is only defined when compiled for __AES__. So anyplace you call it will also need to be in `#ifdef __AES__` and have some non-AES alternative.

Reply:
That's right - this particular hash only makes sense when the processor supports those AES instructions. My guess was it would be reserved for people who know they are on recent enough x86 platforms.

I'm kind of on the fence about including it to be honest. Its faster, but only if you have a fairly large buffer. Given that OIIO has to be portable, its unlikely we'd use it internally.

Reply:
I'd be very comfortable with it as long as the `#if` is *inside* the function -- in other words, if it has an AES-less fallback.